aexpect/ops.py

```diff
@@ -0,0 +1,479 @@
+# This Python file uses the following encoding: utf-8
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Intra2net AG and aexpect contributors
+
+"""
+A list of frequent operations performed through remote sessions.
+
+Most commands passed to a session are routine invocations of common userland
+tools. Nevertheless some of them require a certain effort to set up because
+they are squeezed through the eye of the needle, the shell session.
+
+The functionality in this module assumes a guest with the ordinary userland
+tools available. Some of it might just work on Windows guests if the proper
+environment (i. e. Cygwin) is present.
+
+This module aims to provide convenient wrappers for butter-and-bread
+functionality on a Linux system. Currently, the implemented functions fall
+under four categories: ``grep``, ``stat``, ``test``, and common file ops.
+
+grep:
+Standard case search operations. It is always assumed that binaries in the path
+should be searched (the ``-a`` flag).
+
+stat:
+Allows passing an optional format argument (``-c``). Custom wrappers for
+extracting *atime*, *ctime*, and *mtime*, etc exist.
+
+test:
+Allows for testing existence, permissions and other properties of files.
+
+file ops:
+Simple file operations executed through sessions. Most functions boil down to
+executing simple commands like ls, tar, md5sum, etc. If these operations return
+non-0, a :py:class:`RuntimeError` is raised, containing the commands's error
+message. All functions :py:func:`shlex.quote` their args for better security.
+
+"""
+
+import re
+from os.path import join
+from os.path import split as path_split
+from shlex import quote
+
+from aexpect.exceptions import ShellCmdError
+
+# Need this import for sphinx and other documentation to produce links later on
+# from .client import ShellSession
+
+
+###############################################################################
+# grep(1)
+###############################################################################
+
+
+def grep(session, expr, path, check=False, flags=None):
+    """
+    Invoke ``grep`` on guest searching for *expr* in *path*. Throws a
+    ``TypeError`` in case the API assumptions are violated.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str expr: search expression
+    :param str path: file to search
+    :param bool check: whether to quietly run grep for a boolean check
+    :param str flags: extra flags passed to ``grep`` on the command line
```

Comment:
How much deep do you want to go in abstracting here? I'm usually in favour of simplicity, rather than complex logic and high level default usage with unmodified low-level tweaks if needed.

In this case we have high level API `grep(session, expr, path)`. Optionally we can check status, which is still high-level. Good so far. Then we have `flags`, which is something lower-level and it is specified as `string` and is processed by automatic logic to prefix each one with `-`. This works well with single-letter options like `oP` pr `q`, but does not really fit the long arguments (`--line-regexp`). One could try using lists, which would indeed iterate through strings, but it would silently ignore the option as `-` is not alpha-numeric.

Personally I'd prefer no modifications at all, requiring the users to pass the lower-level arguments themselves, or full-blown API that would take care of translating options (probably list of well defined supported options). This seems like a half-way there and it limits the usage.

Note: I used a long text description here as I see similar issues elsewhere and I'd like to use this first occurrence to properly explain my view and negotiate some agreement on a style for the whole module.

Reply:
Note that one option would be to specify the high-level API by keywords (like `expr`, `path`, `check`) and add `extra_params` to each command, that would be passed directly to the command with no chacks at all, allowing advanced tweaking of that command and would be common in all commands (that consist of a single call).

Reply:
I agree, I will use `flags` here and unify the approach with a few other locations. I could then add a more universal `flags` argument once I refactor some of the internal command handling in a follow up pull request.

Reply:
Sure, can be added later although try not to break the existing API.